kube/src/config/file_config.rs

```diff
@@ -140,15 +142,168 @@ impl Kubeconfig {
+                }
+            }
+        }
         Ok(config)
     }
 
-    /// Read a Config from the default location
+    /// Read a Config from `KUBECONFIG` or the the default location.
     pub fn read() -> Result<Kubeconfig> {
-        let path = utils::find_kubeconfig()?;
-        Self::read_from(path)
+        match Self::from_env()? {
+            Some(config) => Ok(config),
+            None => {
+                let path = utils::default_kube_path().ok_or(ConfigError::NoKubeconfigPath)?;
+                Self::read_from(path)
+            }
+        }
+    }
+
+    /// Create `Kubeconfig` from `KUBECONFIG` environment variable.
+    /// Supports list of files to be merged. `KUBECONFIG` value must contain a valid unicode data.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `KUBECONFIG` value contains the NUL character.
+    pub fn from_env() -> Result<Option<Self>> {
+        match std::env::var(KUBECONFIG) {
+            Ok(value) => {
+                // REVIEW Does Windows require splitting by ';'?
+                let paths = value.split(':').filter(|s| !s.is_empty()).collect::<Vec<_>>();
+                if paths.is_empty() {
+                    return Ok(None);
+                }
+
+                let merged = paths.iter().try_fold(Kubeconfig::default(), |m, p| {
+                    Kubeconfig::read_from(p).and_then(|c| m.merge(c))
+                })?;
+                Ok(Some(merged))
+            }
+
+            Err(std::env::VarError::NotPresent) => Ok(None),
+            Err(std::env::VarError::NotUnicode(_)) => Err(ConfigError::InvalidKubeconfigEnv.into()),
+        }
+    }
+
+    /// Merge kubeconfig file according to the rules described in
+    /// <https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files>
+    ///
+    /// > Merge the files listed in the `KUBECONFIG` environment variable according to these rules:
+    /// >
+    /// > - Ignore empty filenames.
+    /// > - Produce errors for files with content that cannot be deserialized.
+    /// > - The first file to set a particular value or map key wins.
+    /// > - Never change the value or map key.
+    /// >   Example: Preserve the context of the first file to set `current-context`.
+    /// >   Example: If two files specify a `red-user`, use only values from the first file's `red-user`.
+    /// >            Even if the second file has non-conflicting entries under `red-user`, discard them.
+    fn merge(mut self, mut next: Kubeconfig) -> Result<Self> {
+        if self.kind.is_some() && next.kind.is_some() && self.kind != next.kind {
+            return Err(ConfigError::KindMismatch.into());
+        }
+        if self.api_version.is_some() && next.api_version.is_some() && self.api_version != next.api_version {
+            return Err(ConfigError::ApiVersionMismatch.into());
+        }
+
+        if self.kind.is_none() {
+            self.kind = next.kind.take();
+        }
```

Comment:
Oh, it can just be
```rust
self.kind = self.kind.or(next.kind);
```
:laughing: I don't know why I didn't think of that.

`.ok_or_else` needs `.ok`
```rust
self.kind = self.kind.ok_or_else(|| next.kind.take()).ok();
```